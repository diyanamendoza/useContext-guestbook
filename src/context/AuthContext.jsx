import { createContext, useState, useContext } from "react";

const AuthContext = createContext()

const AuthProvider = ({ children }) => {
    const [user, setUser] = useState('')
    const login = (username, password) => {
        const loginOk = username === process.env.AUTH_USER && password === process.env.AUTH_PASS
        if (loginOk) setUser({ username })
        return loginOk
    }
    const logout = () => {
        setUser(null)
    }

    return (
        <AuthContext.Provider value={{ user, login, logout }}> {children} </AuthContext.Provider>
    )
}

const useAuth = () => {
    const context = useContext(AuthContext)
    if (context === undefined) {
        throw new Error('useAuth not used inside provider')
    }
    return context
}

export { AuthProvider, useAuth }