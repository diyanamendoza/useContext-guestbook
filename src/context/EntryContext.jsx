import { useContext, createContext, useState } from 'react'

const EntryContext = createContext()

const EntryProvider = ({ children }) => {
    const [entry, setEntries] = useState([])
    return <EntryContext.Provider value={{ entry, setEntries }}>{children}</EntryContext.Provider>
}

const useEntries = () => {
    const context = useContext(EntryContext)
    if (context === undefined) {
        throw new Error('The useEntries hook is not being called inside a EntryContext.Provider')
    }
    return context
}

export { EntryProvider, useEntries }