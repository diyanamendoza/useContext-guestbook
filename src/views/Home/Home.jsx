import EntryList from "../../components/EntryList/EntryList"
import GuestBook from "../../components/GuestBook/GuestBook"

export default function Home() {
    return (
        <div>
           <GuestBook /> 
           <EntryList />
        </div>
    )
}
