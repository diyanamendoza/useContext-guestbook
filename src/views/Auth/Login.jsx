import { useLocation, useHistory } from 'react-router-dom'
import { useState } from 'react'
import { useAuth } from '../../context/AuthContext'
import styles from '../../App.css'

export default function Login() {
    const history = useHistory()
    const location = useLocation()
    const { from } = location.state || { from: { pathname: '/' }}
    const auth = useAuth()
    const [username, setUsername] = useState('')
    const [pass, setPass] = useState('')
    const [error, setError] = useState(null)

    const handleSubmit = async (e) => {
        e.preventDefault()
        const loginOk = await auth.login(username, pass)
        loginOk ? history.replace(from.pathname) : setError('Login failed.')
    }

    return (
        <div className={styles.formcontainer}>
            <fieldset>
            <legend>Sign In</legend>
                <form onSubmit={handleSubmit}>
                    <label htmlFor="username">Username</label>
                        <input
                        required
                        id="username"
                        type="text"
                        name="username"
                        value={username}
                        onChange={({ target }) => setUsername(target.value)}
                        />
                    <label htmlFor="password">Password</label>
                        <input
                        required
                        id="password"
                        type="password"
                        name="password"
                        value={pass}
                        onChange={({ target }) => setPass(target.value)}
                        />
                    <button type="submit" aria-label="Sign In">Sign In</button>
                </form>
            {error && <p>{error}</p>}
            </fieldset>
        </div>
    )
}
