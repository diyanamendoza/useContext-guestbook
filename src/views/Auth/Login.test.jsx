import {
    fireEvent,
    render,
    screen,
    waitFor,
} from '@testing-library/react'

import { MemoryRouter, Route, Switch } from 'react-router-dom'
import { AuthProvider, useAuth } from '../../context/AuthContext'
import Login from './Login'
import Home from '../Home/Home'
import { EntryProvider } from '../../context/EntryContext'

jest.mock('../../context/AuthContext', () => {
    const originalModule = jest.requireActual('../../context/AuthContext')

    return {
        ...originalModule,
        useAuth: jest.fn(() => {
            return {
                user: { username: 'Betty' },
                login: () => true
            }
        })
    }
})

it('should allow the user Betty to log in', async () => {
    const { container } = render(
        <AuthProvider>
            <EntryProvider>
            <MemoryRouter initialEntries={['/login']}>
                <Switch>
                    <Route exact path='/login'>
                        <Login />
                    </Route>
                    <Route exact path='/'>
                        <Home />
                    </Route>
                </Switch>
            </MemoryRouter>
            </EntryProvider>
        </AuthProvider>
    )

    expect(container).toMatchSnapshot()

    const usernameField = screen.getByLabelText('Username')
    const passField = screen.getByLabelText('Password')
    const submitBtn = screen.getByRole('button', { name: 'Sign In' })

    fireEvent.change(usernameField, {
        target: { value: 'Betty' }
    })

    fireEvent.change(passField, {
        target: { value: 'Test' }
    })

    expect(usernameField).toHaveValue('Betty')
    expect(passField).toHaveValue('Test')

    fireEvent.click(submitBtn)

    return waitFor(() => {
        screen.getByText('Your Entry')
    })
})