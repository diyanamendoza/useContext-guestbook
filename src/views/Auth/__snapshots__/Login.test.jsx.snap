// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow the user Betty to log in 1`] = `
<div>
   
  <div
    class="formcontainer"
  >
    <fieldset>
      <legend>
        Sign In
      </legend>
      <form>
        <label
          for="username"
        >
          Username
        </label>
        <input
          id="username"
          name="username"
          required=""
          type="text"
          value=""
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <button
          aria-label="Sign In"
          type="submit"
        >
          Sign In
        </button>
      </form>
    </fieldset>
  </div>
   
</div>
`;
